### PR TITLE
Split out platform-independent types into a new winit-core crate

### LINKED ISSUE #2759: Split parts of winit into separate crates
I recently stumbled upon #1298, which feels like a pretty good idea to me. For some of the types in this crate, they could probably be split out into a `winit-types` crate (preferably developed as a separate crate in the `winit` repo) that avoid linking to backends.

It looks like, originally, `winit-types` and `glutin-x11-sym` (for X11 types) were planned to be split into other crates. I agree with `winit-types`, and I can also agree with splitting global X11 resources into another crate, since there are global resources that need a single authoritative source to be managed. The rationale is similar to that of the [`signal-hook-registry`](https://docs.rs/signal-hook-registry/latest/signal_hook_registry/). However,  there's no harm in having several X11 shared libraries lying around following AltF02/x11-rs#165

I think that `winit-types` and a hypothetical `winit-xlib` crate would be good candidates to be split off of `winit`. Is there any other room for splitting here? It feels like anything further would require us to slice and dice the `platform_impl` module, which just sounds like a pain. On top of that, the more we split off, the larger the dependency tree gets, and we have to consider if the gains in compartmentalization are worth it.

Would probably need to wait until after #2662

### PR DESCRIPTION
**Pull request by notgull**
This PR implements the first item of #3433 by splitting the platform-independent types from `winit` into a new crate: `winit-core`. It contains mostly types, traits and structures from `winit` that are platform independent. As part of this commit, I made a handful of changes to certain types (most notably `Event`) to make them more platform-independent.

A brief summary of the changes made:

- `dpi` and `keyboard` modules are copied over wholesale.
- The errors in the `error` module that don't involve `OsError` are copied as well.
- Some types from the `window` module are copied verbatim. `WindowId` is now a newtype over `u64` instead of a platform-specific value. `ActivationToken` is copied and given methods to make it useful.
- `Event` is given a new generic parameter to describe platform-specific key information. `InnerSizeWriter` is given methods to allow for instantiation. Aside from this, most things from the `event` module are copied verbatim.
- `AsyncRequestSerial` and `ControlFlow` are copied from `event_loop`. `AsyncRequestSerial::get` is made public.

Aside from this, the biggest changes are just moving everything to `winit-core` and then re-exporting it.

I've formatted this PR as a series of changes in order to make reviewing easier. @rust-windowing/winit Please review this PR quickly. It will be difficult to rebase on top of new changes and it'll take me the whole day to remake it.

Closes #2759

- [ ] Tested on all platforms changed
  - [x] Ran `cargo check` on all platforms to make sure it builds there.
  - [ ] Did not actually run example programs on all platforms.
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [x] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [x] Created or updated an example program if it would help users understand this functionality
- [x] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

### CODE REVIEW

**Comment by notgull** [comment]
@jackpot51 Any insight into why the Redox CI is failing with `__errno not found` errors?

**Comment by jackpot51** [comment]
@notgull not sure, it may be trying to compile with the host compiler instead of a cross compiler?

**Comment by notgull** [comment]
It appears that the only failures are the Redox failure (is that new?) and what appears to be a bug in `cargo-apk`. Both of which are unrelated to this PR.

**Comment by notgull** [comment]
No, the Redox failure is new. It looks like Redox can't compile `run-wasm`.

**Comment by kchibisov** [comment]
Why it compiles it in the first place?

**Comment by notgull** [comment]
> Why it compiles it in the first place?

Because with the new workspace layout it tries to compile everything instead of just `winit`

File: winit/src/platform_impl/ios/mod.rs [hunk 1]
```diff
@@ -68,35 +68,22 @@ mod window;
...
 };
 pub(crate) use crate::cursor::NoCustomCursor as PlatformCustomCursor;
 pub(crate) use crate::cursor::NoCustomCursor as PlatformCustomCursorBuilder;
 pub(crate) use crate::icon::NoIcon as PlatformIcon;
 pub(crate) use crate::platform_impl::Fullscreen;
 
-/// There is no way to detect which device that performed a certain event in
-/// UIKit (i.e. you can't differentiate between different external keyboards,
-/// or wether it was the main touchscreen, assistive technologies, or some
-/// other pointer device that caused a touch event).
```
Comment (madsmtm) [inline winit/src/platform_impl/ios/mod.rs:89]: I'd like to retain this comment in some shape or form, it is nice to know that device IDs will never be supported on iOS.

File: winit-core/src/event.rs [hunk 2]
```diff
@@ -683,7 +665,7 @@ pub struct RawKeyEvent {
 
 /// Describes a keyboard input targeting a window.
 #[derive(Debug, Clone, Eq, PartialEq, Hash)]
-pub struct KeyEvent {
+pub struct KeyEvent<Extra> {
```
Comment (madsmtm) [inline winit-core/src/event.rs:668]: I dislike this design, and think we need to discuss further how to do this sort of thing. See also https://github.com/rust-windowing/winit/issues/3064.

File: Cargo.toml [hunk 3]
```diff
@@ -1,265 +1,16 @@
...
-    # iOS
-    "x86_64-apple-ios",
-    # Android
-    "aarch64-linux-android",
-    # Web
-    "wasm32-unknown-unknown",
+[workspace]
+members = [
+    "run-wasm",
+    "winit",
```
Comment (madsmtm) [inline Cargo.toml:4]: I think it's a mistake to move `winit` to a subdirectory, at least yet, and it also complicates this PR unnecessarily. There is especially value in having the top-level `examples` directory. Let's use the [`include`](https://doc.rust-lang.org/cargo/reference/manifest.html#the-exclude-and-include-fields) field instead to specify which files should be included in the `winit `package.

**Review by madsmtm — CHANGES_REQUESTED** [review]

**Pull request closed without merging**